File: MobileApp/lib/features/videos/presentation/layouts/videos_layout_data.dart
import 'package:flutter/widgets.dart';
import 'package:mobile_app/features/videos/data/video.dart';
import 'package:mobile_app/features/videos/data/video_helpers.dart';
import 'package:mobile_app/features/videos/data/video_selectors.dart';

/// One tag's videos, in relevancy order.
typedef VideoTagSection = ({String tag, List<Video> videos});

/// Everything a videos layout renders, derived once so all five values
/// are handed the identical payload.
///
/// This is where the arrangement-only invariant is made structural: a
/// layout cannot reach past this object, so it cannot fetch, re-sort,
/// or filter anything. Choosing a format changes where these land, not
/// what they are.
class VideosLayoutData {
  const VideosLayoutData({
    required this.tabLabels,
    required this.selectedTabIndex,
    required this.onTabSelected,
    required this.featured,
    required this.sections,
    required this.onVideoTap,
    required this.onViewAll,
  });

  /// Derives the whole payload from a loaded feed and the active top
  /// filter. `scope == null` is the "All" tab.
  factory VideosLayoutData.fromFeed({
    required List<Video> videos,
    required String? scope,
    required ValueChanged<String?> onScopeSelected,
    required ValueChanged<Video> onVideoTap,
    required ValueChanged<String> onViewAll,
  }) {
    final scopes = bigGroupsInFeed(videos);
    final tabScopes = <String?>[null, ...scopes];
    var index = tabScopes.indexOf(scope);
    if (index < 0) index = 0;
    final active = tabScopes[index];

    return VideosLayoutData(
      tabLabels: ['All', ...scopes.map(displayLabel)],
      selectedTabIndex: index,
      onTabSelected: (i) => onScopeSelected(tabScopes[i]),
      featured: featuredVideo(videos, active),
      sections: tagSections(videos, active),
      onVideoTap: onVideoTap,
      onViewAll: onViewAll,
    );
  }

  /// "All" plus one label per `big_group` the loaded feed carries.
  final List<String> tabLabels;
  final int selectedTabIndex;
  final ValueChanged<int> onTabSelected;

  /// The hero: the most relevant video in the active filter.
  final Video? featured;

  /// One entry per tag present in the active filter.
  final List<VideoTagSection> sections;

  final ValueChanged<Video> onVideoTap;

  /// Opens one tag's full list. Called with the tag's wire value.
  final ValueChanged<String> onViewAll;

  /// The active filter holds nothing to show.
  bool get isEmpty => featured == null && sections.isEmpty;

  /// The title a section renders, formatted from its wire tag.
  String titleOf(VideoTagSection section) => displayLabel(section.tag);
}

File: ThemeService/ThemeReact/src/app/showcase/videos/videosLayoutData.ts
// Ports ../../../../../../MobileApp/lib/features/videos/presentation/layouts/
// videos_layout_data.dart — everything a videos ARRANGEMENT renders, derived
// once so all five values are handed the identical payload.
//
// This is where the arrangement-only invariant is made STRUCTURAL rather than
// asserted: an arrangement is given this object and cannot reach past it, so it
// cannot fetch, re-sort, filter, or invent. Choosing an arrangement changes
// where these land, not what they are.
//
// WHAT DART CARRIES THAT THIS DOES NOT: the four callbacks
// (`onScopeSelected` / `onTabSelected` / `onVideoTap` / `onViewAll`). Every
// surface in this island is a preview inside a phone frame that takes no input
// — ../VideoCategoryTabs.tsx and ../rewards/RewardsTabs.tsx both reached the
// same conclusion — so there is nothing for an arrangement to wire them to. The
// affordances themselves are all still rendered; only the routes behind them
// are absent, which is a property of the browser, not of an arrangement.
//
// WHAT DART CALLS A TAG, THIS CALLS A GENRE. Dart's `VideosLayoutData.fromFeed`
// derives its filter from the coarse `big_groups` of the PRE-PORTAL feed model;
// the live screen this island ported (see ./videoSelectors.ts, "WHICH SELECTORS
// THESE ARE") filters on the fine-grained genre set instead. The shape is
// identical — a label list, a hero, one section per key — so the arrangements
// port unchanged; only the vocabulary underneath differs.

import type { ShowcaseVideo } from '../showcaseContent';

import type { VideoGenreSection } from './videoSelectors';
import { featuredVideo, genreLabel, genresInFeed, genreSections } from './videoSelectors';

/**
 * `tabGenres = [null, ...genres]` — "All" is this screen, so index 0 is always
 * the selected tab in a preview that cannot navigate.
 */
export function videoTabLabels(videos: readonly ShowcaseVideo[]): readonly string[] {
  return ['All', ...genresInFeed(videos).map(genreLabel)];
}

/** `VideosLayoutData` — the payload every arrangement is handed. */
export interface VideosLayoutData {
  /** "All" plus one label per genre the loaded feed carries. */
  readonly tabLabels: readonly string[];
  readonly selectedTabIndex: number;
  /** The hero: the top card of the already-ranked page, or null when empty. */
  readonly featured: ShowcaseVideo | null;
  /** One entry per genre present, in first-appearance order. */
  readonly sections: readonly VideoGenreSection[];
  /** The active filter holds nothing to show. */
  readonly isEmpty: boolean;
}

/** `VideosLayoutData.fromFeed` — derives the whole payload from a loaded feed. */
export function videosLayoutData(videos: readonly ShowcaseVideo[]): VideosLayoutData {
  const featured = featuredVideo(videos);
  const sections = genreSections(videos);
  return {
    tabLabels: videoTabLabels(videos),
    selectedTabIndex: 0,
    featured,
    sections,
    isEmpty: featured === null && sections.length === 0,
  };
}

/** `VideosLayoutData.titleOf` — the title a section renders. */
export function sectionTitle(section: VideoGenreSection): string {
  return genreLabel(section.genre);
}
